internal/controller/gitrepository_controller.go

```diff
@@ -650,13 +651,21 @@ func (r *GitRepositoryReconciler) getAuthOpts(ctx context.Context, obj *sourcev1
 	}
 
 	// Configure provider authentication if specified in spec
-	if obj.GetProvider() == sourcev1.GitProviderAzure {
+	switch obj.GetProvider() {
+	case sourcev1.GitProviderAzure:
 		authOpts.ProviderOpts = &git.ProviderOptions{
-			Name: obj.GetProvider(),
+			Name: sourcev1.GitProviderAzure,
 			AzureOpts: []azure.OptFunc{
 				azure.WithAzureDevOpsScope(),
 			},
 		}
+	case sourcev1.GitProviderGitHub:
```

Comment:
Improved the error messaging for the following 4 scenarios, added unit tests and tested with SC.

1) Provider: `github`, no `secretRef` in `gitRepository` spec
    Result: Stalling error, error message: secretRef with github app data must be specified when provider is set to github
```
Name:         ghapp-repo
Namespace:    default
Labels:       <none>
Annotations:  <none>
API Version:  source.toolkit.fluxcd.io/v1
Kind:         GitRepository
Metadata:
  Creation Timestamp:  2024-12-05T20:24:11Z
  Finalizers:
    finalizers.fluxcd.io
  Generation:        1
  Resource Version:  17094229
  UID:               67a75e2e-55f6-4e60-bb04-187e0698a5bb
Spec:
  Interval:  1m
  Provider:  github
  Ref:
    Branch:  main
  Timeout:   60s
  URL:       https://github.com/dp-flux-testing/dp-flux-testing.git
Status:
  Conditions:
    Last Transition Time:  2024-12-05T20:24:11Z
    Message:               secretRef with github app data must be specified when provider is set to github
    Observed Generation:   1
    Reason:                AuthenticationFailed
    Status:                True
    Type:                  Stalled
    Last Transition Time:  2024-12-05T20:24:11Z
    Message:               secretRef with github app data must be specified when provider is set to github
    Observed Generation:   1
    Reason:                AuthenticationFailed
    Status:                False
    Type:                  Ready
    Last Transition Time:  2024-12-05T20:24:11Z
    Message:               secretRef with github app data must be specified when provider is set to github
    Observed Generation:   1
    Reason:                AuthenticationFailed
    Status:                True
    Type:                  FetchFailed
  Observed Generation:     1
Events:
  Type     Reason                Age   From               Message
  ----     ------                ----  ----               -------
  Warning  AuthenticationFailed  28s   source-controller  secretRef with github app data must be specified when provider is set to github
```

2) provider: `github`, `secretRef` is present, but secret is not present
    Result: Generic error with retry.

```
Name:         ghapp-repo
Namespace:    default
Labels:       <none>
Annotations:  <none>
API Version:  source.toolkit.fluxcd.io/v1
Kind:         GitRepository
Metadata:
  Creation Timestamp:  2024-12-05T20:24:11Z
  Finalizers:
    finalizers.fluxcd.io
  Generation:        2
  Resource Version:  17095637
  UID:               67a75e2e-55f6-4e60-bb04-187e0698a5bb
Spec:
  Interval:  1m
  Provider:  github
  Ref:
    Branch:  main
  Secret Ref:
    Name:   ghapp-secret
  Timeout:  60s
  URL:      https://github.com/dp-flux-testing/dp-flux-testing.git
Status:
  Conditions:
    Last Transition Time:  2024-12-05T20:29:13Z
    Message:               building artifact
    Observed Generation:   2
    Reason:                ProgressingWithRetry
    Status:                True
    Type:                  Reconciling
    Last Transition Time:  2024-12-05T20:29:13Z
    Message:               failed to get secret 'default/ghapp-secret': secrets "ghapp-secret" not found
    Observed Generation:   2
    Reason:                AuthenticationFailed
    Status:                False
    Type:                  Ready
    Last Transition Time:  2024-12-05T20:29:02Z
    Message:               failed to get secret 'default/ghapp-secret': secrets "ghapp-secret" not found
    Observed Generation:   2
    Reason:                AuthenticationFailed
    Status:                True
    Type:                  FetchFailed
  Observed Generation:     1
Events:
  Type     Reason                Age               From               Message
  ----     ------                ----              ----               -------
  Warning  AuthenticationFailed  5m9s              source-controller  secretRef with github app data must be specified when provider is set to github
  Warning  AuthenticationFailed  7s (x5 over 18s)  source-controller  failed to get secret 'default/ghapp-secret': secrets "ghapp-secret" not found
```

3) provider: `github`, associated secret has missing app ID/app Installation ID/private key.
    Result: Generic error, validation in pkg code during checkout. 
    
```
Name:         ghapp-repo
Namespace:    default
Labels:       <none>
Annotations:  <none>
API Version:  source.toolkit.fluxcd.io/v1
Kind:         GitRepository
Metadata:
  Creation Timestamp:  2024-12-05T20:24:11Z
  Finalizers:
    finalizers.fluxcd.io
  Generation:        2
  Resource Version:  17097352
  UID:               67a75e2e-55f6-4e60-bb04-187e0698a5bb
Spec:
  Interval:  1m
  Provider:  github
  Ref:
    Branch:  main
  Secret Ref:
    Name:   ghapp-secret
  Timeout:  60s
  URL:      https://github.com/dp-flux-testing/dp-flux-testing.git
Status:
  Conditions:
    Last Transition Time:  2024-12-05T20:35:25Z
    Message:               building artifact
    Observed Generation:   2
    Reason:                ProgressingWithRetry
    Status:                True
    Type:                  Reconciling
    Last Transition Time:  2024-12-05T20:35:25Z
    Message:               failed to checkout and determine revision: private key must be provided to use github app authentication
    Observed Generation:   2
    Reason:                GitOperationFailed
    Status:                False
    Type:                  Ready
    Last Transition Time:  2024-12-05T20:35:25Z
    Message:               failed to checkout and determine revision: private key must be provided to use github app authentication
    Observed Generation:   2
    Reason:                GitOperationFailed
    Status:                True
    Type:                  FetchFailed
  Observed Generation:     1
Events:
  Type     Reason                Age                    From               Message
  ----     ------                ----                   ----               -------
  Warning  AuthenticationFailed  13m                    source-controller  secretRef with github app data must be specified when provider is set to github
  Warning  AuthenticationFailed  7m43s (x7 over 8m30s)  source-controller  failed to get secret 'default/ghapp-secret': secrets "ghapp-secret" not found
  Warning  GitOperationFailed    6m55s                  source-controller  failed to checkout and determine revision: app ID must be provided to use github app authentication
  Warning  GitOperationFailed    5m19s                  source-controller  failed to checkout and determine revision: app installation ID must be provided to use github app authentication
  Warning  GitOperationFailed    2m7s                   source-controller  failed to checkout and determine revision: private key must be provided to use github app authentication
```

4) secret contains githubAppID and provider not set. 
    Result: Stalling error indicating provider is not set to github

Notes:
-  As discussed in the dev meeting, checking only for github App ID presence to detect that intent is to use github app authentication. 
- Did not include hostname validation here since for Github Enterprise Server, the hostname does not necessarily contain 'github'. In any case the `secret` should not contain githubAppID as data unless intent was to use github app authentication. Let me know thoughts. 

```
Name:         ghapp-repo
Namespace:    default
Labels:       <none>
Annotations:  <none>
API Version:  source.toolkit.fluxcd.io/v1
Kind:         GitRepository
Metadata:
  Creation Timestamp:  2024-12-05T20:24:11Z
  Finalizers:
    finalizers.fluxcd.io
  Generation:        3
  Resource Version:  17099528
  UID:               67a75e2e-55f6-4e60-bb04-187e0698a5bb
Spec:
  Interval:  1m
  Ref:
    Branch:  main
  Secret Ref:
    Name:   ghapp-secret
  Timeout:  60s
  URL:      https://github.com/dp-flux-testing/dp-flux-testing.git
Status:
  Artifact:
    Digest:            sha256:312bafaed0040134c57615608de187def57d97d4db09e46a3d64562b6f9c96ed
    Last Update Time:  2024-12-05T20:41:50Z
    Path:              gitrepository/default/ghapp-repo/f3e6c65b2ae4e0f333cf317143b20efcc83400a0.tar.gz
    Revision:          main@sha1:f3e6c65b2ae4e0f333cf317143b20efcc83400a0
    Size:              470
    URL:               http://source-controller.source-system.svc.cluster.local./gitrepository/default/ghapp-repo/f3e6c65b2ae4e0f333cf317143b20efcc83400a0.tar.gz
  Conditions:
    Last Transition Time:  2024-12-05T20:43:18Z
    Message:               secretRef 'default/ghapp-secret' has github app data but provider is not set to github
    Observed Generation:   3
    Reason:                AuthenticationFailed
    Status:                True
    Type:                  Stalled
    Last Transition Time:  2024-12-05T20:43:18Z
    Message:               secretRef 'default/ghapp-secret' has github app data but provider is not set to github
    Observed Generation:   3
    Reason:                AuthenticationFailed
    Status:                False
    Type:                  Ready
    Last Transition Time:  2024-12-05T20:43:18Z
    Message:               secretRef 'default/ghapp-secret' has github app data but provider is not set to github
    Observed Generation:   3
    Reason:                AuthenticationFailed
    Status:                True
    Type:                  FetchFailed
    Last Transition Time:  2024-12-05T20:41:50Z
    Message:               stored artifact for revision 'main@sha1:f3e6c65b2ae4e0f333cf317143b20efcc83400a0'
    Observed Generation:   2
    Reason:                Succeeded
    Status:                True
    Type:                  ArtifactInStorage
  Observed Generation:     3
```